### DISCUSSION TITLE
How to handle Thing Models?

### DISCUSSION THREAD

**Discussion by bachp** [sifis-home/wot-td#112] [General]
I'm using wot-td to parse Thing Descriptions. This works well.

As [Thing Models](https://www.w3.org/TR/wot-thing-description11/#introduction-tm) also seem to be thing descriptions I tried to use the same mechanism to parse them. This, however, failed as they don't fully comply with the schema of a TD used in wot-td.

The reason is they might have certain fields missing or using [placeholders](https://www.w3.org/TR/wot-thing-description11/#thing-model-td-placeholder) instead of real values.

Now for me the question is how should Thing models be supported by the library?

I see several possibilities:

1. Make the current Thing type laxer to also support Thing Models
2. Add a new `ThingModel` type. In the future we could provide functionality that allows transforming a `ThingModel` to a concrete Thing. Maybe even in a safe why that requires the user to supply all the missing information.
3. In a separate crate.

Feedback would be appreciated.

**Comment by lu-zero** [comment]
A separate `Model` struct would be the best, you can recycle all the structures but the actual `Thing` and then add the placeholders. `Thing` itself is already quite complex so I wouldn't add on top of it.

**Reply by bachp** [reply]
That was also my intuition. I will give it a try and come up with an PR if I'm successful.

**Reply by lu-zero** [reply]
I'm looking forward to it :) You can reach us on discord or telegram if you need directions :)
